<h1>2nd Contribution</h1>

<h3>Feature Selection Process</h3></br>
<p>The following statistical tools is used to filter and select:</p>
<u1><li>Variance</li>
<li>Correlation</li>
    <li>Association strength</li>
</u1>

<h3>Variance </h3></br>
<p>Variance was used to determine features that would be dropped, as features with variance closer to zero concludes it's low likelihood of being useful in terms of predictability <strong></strong>and can only be used on numerical columns. 
    Note: Values would not be dropped based of variance alone as it might have a high correlation and can still be vital tool to decisions making</p>

In [14]:
# Variance check
low_variance_list=[]
low_variance_list2=[]
position=0
# Variance check of numerical data
y=[continous_features.var()]
print(y)

print("\nVariance of all the numerical data in full representation\n"+"__________________________________________________\n")
for i in (continous_features.var()):
    print (str(position)+":  "+str(i))
    if i<1:
#         print("Has a low variance is not good for prediction")
        low_variance_list.append(position)
    position+=1


print("\n\n"+"List of values with low variance amongst the numerical data\n"
      +"_____________________________________________\n\n"+str(low_variance_list)+"\n\n")
# I have created a list of all the features and their variances
low_variance_list2=str(continous_features.var()).split('\n')

# print(continous_features.var())
# Get all the low variance features
for i in low_variance_list:
    print(low_variance_list2[i])
    
# Below is the list of values with low variance

[LotFrontage      4.850576e+02
LotArea          9.962565e+07
MasVnrArea       3.260520e+04
BsmtFinSF1       2.080255e+05
BsmtFinSF2       2.602391e+04
BsmtUnfSF        1.952464e+05
TotalBsmtSF      1.924624e+05
1stFlrSF         1.494501e+05
2ndFlrSF         1.905571e+05
LowQualFinSF     2.364204e+03
GrLivArea        2.761296e+05
BsmtFullBath     2.692682e-01
BsmtHalfBath     5.700283e-02
FullBath         3.035082e-01
HalfBath         2.528937e-01
BedroomAbvGr     6.654938e-01
KitchenAbvGr     4.854892e-02
TotRmsAbvGrd     2.641903e+00
Fireplaces       4.155947e-01
GarageCars       5.584797e-01
GarageArea       4.571251e+04
WoodDeckSF       1.570981e+04
OpenPorchSF      4.389861e+03
EnclosedPorch    3.735550e+03
3SsnPorch        8.595059e+02
ScreenPorch      3.108889e+03
PoolArea         1.614216e+03
MiscVal          2.461381e+05
SalePrice        6.311111e+09
dtype: float64]

Variance of all the numerical data in full representation
__________________________________________________

0:

<p>The above list of shows the features with low variance</p>

<h3>Correlation Check</h3>

</br>
<h4>Numerical values</h4>
<p>Features between the values (-0.3 and +0.3) are considered to have a <strong>Weak correlation</strong>.
    Obervation will be noted for the features in the range <strong>greater than -0.3 and less than +0.3</strong> </p>

In [15]:
correlation_values=continous_features[continous_features.columns[1:]].corr()['SalePrice'][:]
print(""+str(correlation_values))



indexes=[]
position=0
for i in correlation_values:
#     print(i)
    if i>(-0.3) and i<0.3:    
        indexes.append(position)
    position+=1
    


print("\nThe positions of the columns with the weakest correlation\n"
      "_____________________________________________________\n\n"+str(indexes)+"\n")
for i in indexes:
    print(correlation_values[i])
        


LotArea          0.263843
MasVnrArea       0.475241
BsmtFinSF1       0.386420
BsmtFinSF2      -0.011378
BsmtUnfSF        0.214479
TotalBsmtSF      0.613581
1stFlrSF         0.605852
2ndFlrSF         0.319334
LowQualFinSF    -0.025606
GrLivArea        0.708624
BsmtFullBath     0.227122
BsmtHalfBath    -0.016844
FullBath         0.560664
HalfBath         0.284108
BedroomAbvGr     0.168213
KitchenAbvGr    -0.135907
TotRmsAbvGrd     0.533723
Fireplaces       0.466929
GarageCars       0.640409
GarageArea       0.623431
WoodDeckSF       0.324413
OpenPorchSF      0.315856
EnclosedPorch   -0.128578
3SsnPorch        0.044584
ScreenPorch      0.111447
PoolArea         0.092404
MiscVal         -0.021190
SalePrice        1.000000
Name: SalePrice, dtype: float64

The positions of the columns with the weakest correlation
_____________________________________________________

[0, 3, 4, 8, 10, 11, 13, 14, 15, 22, 23, 24, 25, 26]

0.2638433538714051
-0.011378121450215146
0.21447910554696928
-0.02560613

<h4>Conclusion on correlation </h4>
</br><p>Above is the list of compiled variables with a weak correlation</p>

In [16]:
print(continous_features.iloc[:,0])
names=[]
to_drop=[]
for col in continous_features.columns: names.append(col)
    
for i in indexes:
    if i in low_variance_list:
        to_drop.append(names[i])

print("\nThe Columns to be dropped\n"+"___________________________\n\n"+str(to_drop))


for i in to_drop:
    continous_features=continous_features.drop([i], axis=1)
    
print("The columns have been dropped from the continous dataset")
continous_features

0       65.0
1       80.0
2       68.0
3       60.0
4       84.0
        ... 
1455    62.0
1456    85.0
1457    66.0
1458    68.0
1459    75.0
Name: LotFrontage, Length: 1460, dtype: float64

The Columns to be dropped
___________________________

['BsmtFullBath', 'FullBath', 'HalfBath', 'BedroomAbvGr']
The columns have been dropped from the continous dataset


LotFrontage  LotArea  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
0            65.0     8450       196.0         706           0        150   
1            80.0     9600         0.0         978           0        284   
2            68.0    11250       162.0         486           0        434   
3            60.0     9550         0.0         216           0        540   
4            84.0    14260       350.0         655           0        490   
...           ...      ...         ...         ...         ...        ...   
1455         62.0     7917         0.0           0           0        953   
1456         85.0    13175       119.0         790         163        589   
1457         66.0     9042         0.0         275           0        877   
1458         68.0     9717         0.0          49        1029          0   
1459         75.0     9937         0.0         830         290        136   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  ...  GarageCars  \
0             856       856       854             0  ...           2   
1            1262      1262         0             0  ...           2   
2             920       920       866             0  ...           2   
3             756       961       756             0  ...           3   
4            1145      1145      1053             0  ...           3   
...           ...       ...       ...           ...  ...         ...   
1455          953       953       694             0  ...           2   
1456         1542      2073         0             0  ...           2   
1457         1152      1188      1152             0  ...           1   
1458         1078      1078         0             0  ...           1   
1459         1256      1256         0             0  ...           1   

      GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  \
0            548           0           61              0          0   
1            460         298            0              0          0   
2            608           0           42              0          0   
3            642           0           35            272          0   
4            836         192           84              0          0   
...          ...         ...          ...            ...        ...   
1455         460           0           40              0          0   
1456         500         349            0              0          0   
1457         252           0           60              0          0   
1458         240         366            0            112          0   
1459         276         736           68              0          0   

      ScreenPorch  PoolArea  MiscVal  SalePrice  
0               0         0        0     208500  
1               0         0        0     181500  
2               0         0        0     223500  
3               0         0        0     140000  
4               0         0        0     250000  
...           ...       ...      ...        ...  
1455            0         0        0     175000  
1456            0         0        0     210000  
1457            0         0     2500     266500  
1458            0         0        0     142125  
1459            0         0        0     147500  

[1460 rows x 25 columns]

<h3>Conclusion on Numerical Values</h3>
<p>You can see the above list of columns with very little variance and weak correlation to the SalePrice from the said data I have resorted to dropping them 'BsmtFullBath', 'FullBath', 'HalfBath', 'BedroomAbvGr' </p>

<h4>Categorical values</h4></br>
<p>Columns with weak Eta squared score will be dropped as the Eta score shows the associative strength between categorical/non-numerical data and a continous data</p>

In [17]:
to_drop_cat=[]
selected_cat_features={}
for i in correlation_categories:
    if correlation_categories[i]>0.14:
        selected_cat_features[i]=correlation_categories[i]
    elif correlation_categories[i]<0.14:
        to_drop_cat.append(i)

# print(correlation_categories)
# print(selected_cat_features)
print("To be dropped\n\n"+str(to_drop_cat))

# To check and drop the ones from the dataset
for i in to_drop_cat:
#     print(i)
    categorical_features=categorical_features.drop([i], axis=1)

    
categorical_features

# These categories have a medium to high association strength to the SalePricecx

To be dropped

['SaleType', 'SaleCondition', 'OverallCond', 'MSZoning', 'HouseStyle', 'GarageQual', 'GarageCond', 'LotShape', 'CentralAir', 'Electrical', 'RoofStyle', 'PavedDrive', 'BsmtCond', 'Fence', 'BldgType', 'Condition1', 'RoofMatl', 'BsmtFinType2', 'LandContour', 'ExterCond', 'PoolQC', 'LotConfig', 'Alley', 'Functional', 'Heating', 'Condition2', 'MoSold', 'MiscFeature', 'LandSlope', 'YrSold', 'Street', 'Utilities']


LotFrontage  LotArea Neighborhood  OverallQual  YearBuilt  YearRemodAdd  \
0            65.0     8450      CollgCr            7       2003          2003   
1            80.0     9600      Veenker            6       1976          1976   
2            68.0    11250      CollgCr            7       2001          2002   
3            60.0     9550      Crawfor            7       1915          1970   
4            84.0    14260      NoRidge            8       2000          2000   
...           ...      ...          ...          ...        ...           ...   
1455         62.0     7917      Gilbert            6       1999          2000   
1456         85.0    13175       NWAmes            6       1978          1988   
1457         66.0     9042      Crawfor            7       1941          2006   
1458         68.0     9717        NAmes            5       1950          1996   
1459         75.0     9937      Edwards            5       1965          1965   

     Exterior1st Exterior2nd MasVnrType  MasVnrArea  ... GarageCars  \
0        VinylSd     VinylSd    BrkFace       196.0  ...          2   
1        MetalSd     MetalSd       None         0.0  ...          2   
2        VinylSd     VinylSd    BrkFace       162.0  ...          2   
3        Wd Sdng     Wd Shng       None         0.0  ...          3   
4        VinylSd     VinylSd    BrkFace       350.0  ...          3   
...          ...         ...        ...         ...  ...        ...   
1455     VinylSd     VinylSd       None         0.0  ...          2   
1456     Plywood     Plywood      Stone       119.0  ...          2   
1457     CemntBd     CmentBd       None         0.0  ...          1   
1458     MetalSd     MetalSd       None         0.0  ...          1   
1459     HdBoard     HdBoard       None         0.0  ...          1   

     GarageArea WoodDeckSF OpenPorchSF EnclosedPorch  3SsnPorch  ScreenPorch  \
0           548          0          61             0          0            0   
1           460        298           0             0          0            0   
2           608          0          42             0          0            0   
3           642          0          35           272          0            0   
4           836        192          84             0          0            0   
...         ...        ...         ...           ...        ...          ...   
1455        460          0          40             0          0            0   
1456        500        349           0             0          0            0   
1457        252          0          60             0          0            0   
1458        240        366           0           112          0            0   
1459        276        736          68             0          0            0   

      PoolArea  MiscVal SalePrice  
0            0        0    208500  
1            0        0    181500  
2            0        0    223500  
3            0        0    140000  
4            0        0    250000  
...        ...      ...       ...  
1455         0        0    175000  
1456         0        0    210000  
1457         0     2500    266500  
1458         0        0    142125  
1459         0        0    147500  

[1460 rows x 47 columns]

<p>As shown above anything with a weak Eta squared preformance score was dropped from the datafile</p>

In [18]:
# Compiling the two into a single dataset
# I am compiliing the datasets of the different categories of data into a single dataset
# Categorical features data set and Continous features dataset
training_features=pd.merge(categorical_features,continous_features)
training_features

LotFrontage  LotArea Neighborhood  OverallQual  YearBuilt  YearRemodAdd  \
0            65.0     8450      CollgCr            7       2003          2003   
1            80.0     9600      Veenker            6       1976          1976   
2            68.0    11250      CollgCr            7       2001          2002   
3            60.0     9550      Crawfor            7       1915          1970   
4            84.0    14260      NoRidge            8       2000          2000   
...           ...      ...          ...          ...        ...           ...   
1457         62.0     7917      Gilbert            6       1999          2000   
1458         85.0    13175       NWAmes            6       1978          1988   
1459         66.0     9042      Crawfor            7       1941          2006   
1460         68.0     9717        NAmes            5       1950          1996   
1461         75.0     9937      Edwards            5       1965          1965   

     Exterior1st Exterior2nd MasVnrType  MasVnrArea  ... GarageCars  \
0        VinylSd     VinylSd    BrkFace       196.0  ...          2   
1        MetalSd     MetalSd       None         0.0  ...          2   
2        VinylSd     VinylSd    BrkFace       162.0  ...          2   
3        Wd Sdng     Wd Shng       None         0.0  ...          3   
4        VinylSd     VinylSd    BrkFace       350.0  ...          3   
...          ...         ...        ...         ...  ...        ...   
1457     VinylSd     VinylSd       None         0.0  ...          2   
1458     Plywood     Plywood      Stone       119.0  ...          2   
1459     CemntBd     CmentBd       None         0.0  ...          1   
1460     MetalSd     MetalSd       None         0.0  ...          1   
1461     HdBoard     HdBoard       None         0.0  ...          1   

     GarageArea WoodDeckSF OpenPorchSF EnclosedPorch  3SsnPorch  ScreenPorch  \
0           548          0          61             0          0            0   
1           460        298           0             0          0            0   
2           608          0          42             0          0            0   
3           642          0          35           272          0            0   
4           836        192          84             0          0            0   
...         ...        ...         ...           ...        ...          ...   
1457        460          0          40             0          0            0   
1458        500        349           0             0          0            0   
1459        252          0          60             0          0            0   
1460        240        366           0           112          0            0   
1461        276        736          68             0          0            0   

      PoolArea  MiscVal SalePrice  
0            0        0    208500  
1            0        0    181500  
2            0        0    223500  
3            0        0    140000  
4            0        0    250000  
...        ...      ...       ...  
1457         0        0    175000  
1458         0        0    210000  
1459         0     2500    266500  
1460         0        0    142125  
1461         0        0    147500  

[1462 rows x 47 columns]

<h4>Split the set to training and testing data</h4>

In [19]:
# Importing the necessary modules 
from sklearn.model_selection import train_test_split


# Get all the categorical features' column names into a list
cat_features=[]
for col in categorical_features.columns:
    cat_features.append(col)


# To drop the target prediction value
target_value=training_features.iloc[:,-1]
training_features=training_features.drop("SalePrice", axis=1)
# Splitting the train and test dataset to 50%-50%
# random state is to ensure my data siez stays the same
x_train, x_test, y_train, y_test=train_test_split(training_features,target_value, test_size=0.1, random_state=10)



In [20]:
# This was to extract a list from the dictionary that held the categorical data 
# so I could use it in the Hot encoder 
select_cat_list=[]
for i in selected_cat_features:
    select_cat_list.append(i)
print(select_cat_list)

['OverallQual', 'Neighborhood', 'ExterQual', 'BsmtQual', 'KitchenQual', 'YearBuilt', 'GarageYrBlt', 'YearRemodAdd', 'GarageFinish', 'FireplaceQu', 'Foundation', 'GarageType', 'BsmtFinType1', 'HeatingQC', 'MasVnrType', 'Exterior2nd', 'Exterior1st', 'BsmtExposure']


In [35]:
# Training the Model on the feature selected data
# Using linear regression as it is the best in this case
from sklearn.linear_model import LinearRegression 
 
# call linear regression model
lrg=LinearRegression()
# have a change in variable name
trainle, testle=x_train,x_test


trainle


LotFrontage  LotArea Neighborhood  OverallQual  YearBuilt  YearRemodAdd  \
1342    70.000000     8294        NAmes            4       1971          1971   
524     50.000000     5000      BrkSide            6       1947          1950   
1398    70.049958    57200       Timber            5       1948          1950   
1178    37.000000     6951      Mitchel            5       1984          1985   
1245   107.000000    13891      NridgHt           10       2006          2006   
...           ...      ...          ...          ...        ...           ...   
1393    65.000000     8944        NAmes            5       1967          1967   
1344    70.049958     9375      CollgCr            8       2002          2002   
527     62.000000     7500      Somerst            7       2005          2005   
1149    75.000000    12000      Crawfor            7       1941          1950   
1289    70.049958    36500      ClearCr            5       1964          1964   

     Exterior1st Exterior2nd    MasVnrType  MasVnrArea  ... GarageFinish  \
1342     MetalSd     MetalSd          None    0.000000  ...          Unf   
524      CemntBd     CmentBd          None    0.000000  ...          Unf   
1398     Wd Sdng     Wd Sdng          None    0.000000  ...          Unf   
1178     HdBoard     Plywood          None    0.000000  ...          Unf   
1245     VinylSd     VinylSd  value_absent  103.685262  ...          Fin   
...          ...         ...           ...         ...  ...          ...   
1393     Plywood     Plywood          None    0.000000  ...          Unf   
1344     VinylSd     VinylSd       BrkFace  149.000000  ...          RFn   
527      VinylSd     VinylSd          None    0.000000  ...          Unf   
1149     MetalSd     MetalSd          None    0.000000  ...          Unf   
1289     Wd Sdng     Wd Sdng        BrkCmn  621.000000  ...          Unf   

     GarageCars GarageArea WoodDeckSF OpenPorchSF  EnclosedPorch  3SsnPorch  \
1342          4        480          0           0              0          0   
524           2        420          0          24             36          0   
1398          2        572          0           0             50          0   
1178          1        264        362           0              0          0   
1245          3        850        216         229              0          0   
...         ...        ...        ...         ...            ...        ...   
1393          3        792          0         152              0          0   
1344          2        647        192          87              0          0   
527           2        453         38         144              0          0   
1149          1        234          0           0              0          0   
1289          2        390        168         198              0          0   

      ScreenPorch  PoolArea MiscVal  
1342            0         0       0  
524             0         0       0  
1398            0         0       0  
1178            0         0       0  
1245            0         0       0  
...           ...       ...     ...  
1393            0         0       0  
1344            0         0       0  
527             0         0       0  
1149            0         0       0  
1289            0         0       0  

[1315 rows x 46 columns]

***Note***
<p>I used a hot encoder instead of a laber encoder because of the forced ordinality that will lead to the average of two categories leading to another which can be quite a problem</p>

In [36]:
# calling one hot encoder to make dummy columns and drop the excess columns by one to avoid bringing in an error
from sklearn.preprocessing import OneHotEncoder
# The handle unknown argmument makes sure it ignores any new categories it had not seen
ohe=OneHotEncoder(handle_unknown='ignore')

# Using a transofmer I can pick the select columns that are to be One Hot Encoded
from sklearn.compose import make_column_transformer
column_trans=make_column_transformer((ohe,select_cat_list),remainder='passthrough' )


# making a pipeline that will do the following steps o  the data
from sklearn.pipeline import make_pipeline
pipe=make_pipeline(column_trans, lrg)


In [37]:
# training the model
pipe.fit(trainle,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['OverallQual',
                                                   'Neighborhood', 'ExterQual',
                                                   'BsmtQual', 'KitchenQual',
                                                   'YearBuilt', 'GarageYrBlt',
                                                   'YearRemodAdd',
                                                   'GarageFinish',
                                                   'FireplaceQu', 'Foundation',
                                                   'GarageType', 'BsmtFinType1',
                                                   'HeatingQC', 'MasVnrType',
                                                   'Exterior2nd', 'Ex

In [39]:
# Performance of the Test set
pipe.score(testle,y_test)
pipe.score(trainle,y_train)

0.8361199709935456

<h2>Conclusion</h2></br>
<p>As shown in the above analysis our best performance score was extracted after reshaping the data sizes iteratively, resulting into 90% for the Training set and  84% for the test set. Which is not exactly accurate but it is functional when applied</p>